File: App/app/src/main/java/com/cinema/ticket_booking/ui/booking/BookingDetailFragment.java
package com.cinema.ticket_booking.ui.booking;

import android.graphics.Bitmap;
import android.os.Bundle;
import android.view.*;
import com.cinema.ticket_booking.util.SnackbarHelper;
import android.widget.Toast;
import androidx.annotation.*;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;
import com.bumptech.glide.Glide;
import com.cinema.ticket_booking.R;
import com.cinema.ticket_booking.databinding.FragmentBookingDetailBinding;
import com.cinema.ticket_booking.util.TicketCacheManager;
import com.google.zxing.BarcodeFormat;
import com.journeyapps.barcodescanner.BarcodeEncoder;
import dagger.hilt.android.AndroidEntryPoint;
import javax.inject.Inject;

@AndroidEntryPoint
public class BookingDetailFragment extends Fragment {

    private FragmentBookingDetailBinding binding;
    private BookingDetailViewModel viewModel;
    private boolean isCodeVisible = false;
    private String rawBookingCode = "";

    @Inject
    TicketCacheManager ticketCacheManager;

    @Override
    public View onCreateView(@NonNull LayoutInflater i, ViewGroup c, Bundle s) {
        binding = FragmentBookingDetailBinding.inflate(i, c, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        viewModel = new ViewModelProvider(this).get(BookingDetailViewModel.class);

        String bookingId = getArguments() != null ? getArguments().getString("bookingId") : null;
        if (bookingId != null)
            viewModel.loadBooking(bookingId);

        binding.btnBack.setOnClickListener(v -> Navigation.findNavController(view).popBackStack());
        binding.btnBackTop.setOnClickListener(v -> Navigation.findNavController(view).popBackStack());

        binding.ivToggleCode.setOnClickListener(v -> {
            isCodeVisible = !isCodeVisible;
            updateBookingCodeDisplay();
        });

        binding.btnShare.setOnClickListener(v -> {
            if (rawBookingCode.isEmpty()) return;
            android.content.Intent shareIntent = new android.content.Intent(android.content.Intent.ACTION_SEND);
            shareIntent.setType("text/plain");
            shareIntent.putExtra(android.content.Intent.EXTRA_SUBJECT, "Thông tin vé xem phim");
            String shareMessage = "Mình đã đặt vé xem phim *" + binding.tvMovieTitle.getText().toString() + "*\n" +
                                  "Tại: " + binding.tvCinema.getText().toString() + "\n" +
                                  "Lúc: " + binding.tvShowtime.getText().toString() + "\n" +
                                  "Mã đặt vé: " + rawBookingCode;
            shareIntent.putExtra(android.content.Intent.EXTRA_TEXT, shareMessage);
            startActivity(android.content.Intent.createChooser(shareIntent, "Chia sẻ vé"));
        });

        viewModel.getBooking().observe(getViewLifecycleOwner(), resource -> {
            switch (resource.status) {
                case LOADING -> binding.progressBar.setVisibility(View.VISIBLE);
                case SUCCESS -> {
                    binding.progressBar.setVisibility(View.GONE);
                    if (resource.data != null) {
                        renderBooking(resource.data, false);
                    }
                }
                case ERROR -> {
                    binding.progressBar.setVisibility(View.GONE);
                    if (resource.data != null) {
                        renderBooking(resource.data, true);
                        if (resource.message != null && !resource.message.isEmpty()) {
                            SnackbarHelper.showSuccess(binding.getRoot(), resource.message);
                        }
                    } else {
                        SnackbarHelper.showError(binding.getRoot(), resource.message);
                    }
                }
            }
        });
    }

    private void renderBooking(com.cinema.ticket_booking.data.model.response.BookingResponse b, boolean isOffline) {
        rawBookingCode = b.getBookingCode();
        updateBookingCodeDisplay();
        binding.tvMovieTitle.setText(b.getMovieTitle());
        binding.tvCinema.setText("CineNoir " + b.getCinemaName());
        binding.tvScreen.setText(b.getScreenName());
        binding.tvFormat.setText(formatScreenType(b.getScreenType()));
        
        // Format showtime
        String rawTime = b.getStartTime();
        if (rawTime != null) {
            try {
                java.text.SimpleDateFormat sdfIn;
                if (rawTime.contains("T")) {
                    sdfIn = new java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", java.util.Locale.getDefault());
                } else {
                    sdfIn = new java.text.SimpleDateFormat("yyyy-MM-dd HH:mm:ss", java.util.Locale.getDefault());
                }
                java.util.Date d = sdfIn.parse(rawTime);
                java.text.SimpleDateFormat sdfOut = new java.text.SimpleDateFormat("h:mm a - dd/MM/yyyy", java.util.Locale.getDefault());
                binding.tvShowtime.setText(sdfOut.format(d));
            } catch (Exception e) {
                binding.tvShowtime.setText(rawTime);
            }
        } else {
            binding.tvShowtime.setText("");
        }
        
        // Format status
        String statusText = switch (b.getStatus() != null ? b.getStatus() : "") {
            case "PAID" -> "ĐÃ THANH TOÁN";
            case "PENDING" -> "CHỜ THANH TOÁN";
            case "CANCELLED" -> "ĐÃ HỦY";
            case "EXPIRED" -> "HẾT HẠN";
            default -> b.getStatus();
        };

        binding.tvStatus.setText(statusText);
        
        binding.tvTotal.setText(String.format("Tổng thanh toán: %,.0fđ", b.getTotalAmount()));

        // Danh sách ghế
        if (b.getSeats() != null) {
            StringBuilder seats = new StringBuilder();
            for (int i = 0; i < b.getSeats().size(); i++) {
                var s = b.getSeats().get(i);
                seats.append(s.getRowLabel()).append(s.getColNumber());
                if (i < b.getSeats().size() - 1) {
                    seats.append(", ");
                }
            }
            binding.tvSeats.setText(seats.toString().trim());
        }

        // Danh sách bắp nước
        if (b.getCombos() != null && !b.getCombos().isEmpty()) {
            StringBuilder combosStr = new StringBuilder();
            for (int k = 0; k < b.getCombos().size(); k++) {
                var cb = b.getCombos().get(k);
                combosStr.append(cb.getQuantity()).append("x ").append(cb.getComboName());
                if (k < b.getCombos().size() - 1) {
                    combosStr.append("\n");
                }
            }
            binding.tvCombos.setText(combosStr.toString().trim());
            binding.layoutCombos.setVisibility(View.VISIBLE);
        } else {
            binding.layoutCombos.setVisibility(View.GONE);
        }

        // Poster
        if (getContext() != null) {
            Glide.with(this).load(b.getMoviePosterUrl())
                    .placeholder(R.drawable.ic_movie_placeholder).into(binding.ivPoster);
        }

        // QR Code
        if (b.getQrCode() != null && !b.getQrCode().isEmpty()) {
            generateQrCode(b.getQrCode());
        } else {
            binding.ivQrCode.setVisibility(View.GONE);
        }

        // Trạng thái màu
        int statusColor = switch (b.getStatus() != null ? b.getStatus() : "") {
            case "PAID" -> R.color.seat_available;
            case "PENDING" -> R.color.tertiary;
            case "CANCELLED", "EXPIRED" -> R.color.error;
            default -> R.color.on_surface_variant;
        };
        binding.tvStatus.setTextColor(getResources().getColor(statusColor, null));

        if ("PENDING".equals(b.getStatus()) && !isOffline) {
            binding.btnShare.setVisibility(View.GONE);
            binding.btnPay.setVisibility(View.VISIBLE);
            binding.btnPay.setOnClickListener(v -> {
                Bundle args = new Bundle();
                args.putString("bookingId", b.getId());
                if (getView() != null) {
                    Navigation.findNavController(getView()).navigate(R.id.action_bookingDetail_to_payment, args);
                }
            });
        } else {
            binding.btnShare.setVisibility(View.VISIBLE);
            binding.btnPay.setVisibility(View.GONE);
        }
    }

    private void updateBookingCodeDisplay() {
        if (binding == null) return;
        if (isCodeVisible) {
            binding.tvBookingCode.setText(rawBookingCode);
            binding.ivToggleCode.setImageResource(R.drawable.ic_eye);
            binding.tvBookingCode.setLetterSpacing(0.3f);
        } else {
            binding.tvBookingCode.setText("••••••");
            binding.ivToggleCode.setImageResource(R.drawable.ic_eye_off);
            binding.tvBookingCode.setLetterSpacing(0.8f);
        }
    }

    private void generateQrCode(String content) {
        try {
            BarcodeEncoder encoder = new BarcodeEncoder();
            android.graphics.Bitmap bitmap = encoder.encodeBitmap(content, BarcodeFormat.QR_CODE, 400, 400);
            binding.ivQrCode.setImageBitmap(bitmap);
            binding.ivQrCode.setVisibility(View.VISIBLE);
        } catch (Exception e) {
            binding.ivQrCode.setVisibility(View.GONE);
        }
    }

    private String formatScreenType(String rawType) {
        if (rawType == null) return "2D";
        switch (rawType.toUpperCase()) {
            case "STANDARD": return "2D";
            case "THREE_D": return "3D";
            case "IMAX": return "IMAX";
            case "FOUR_DX": return "4DX";
            default: return rawType;
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: App/app/src/main/java/com/cinema/ticket_booking/ui/booking/BookingDetailFragment.kt
package com.cinema.ticket_booking.ui.booking

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.Navigation
import com.bumptech.glide.Glide
import com.cinema.ticket_booking.R
import com.cinema.ticket_booking.data.model.response.BookingResponse
import com.cinema.ticket_booking.databinding.FragmentBookingDetailBinding
import com.cinema.ticket_booking.util.Resource
import com.cinema.ticket_booking.util.SnackbarHelper
import com.cinema.ticket_booking.util.TicketCacheManager
import com.google.android.material.bottomnavigation.BottomNavigationView
import com.google.zxing.BarcodeFormat
import com.journeyapps.barcodescanner.BarcodeEncoder
import dagger.hilt.android.AndroidEntryPoint
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import javax.inject.Inject

@AndroidEntryPoint
class BookingDetailFragment : Fragment() {

    private var _binding: FragmentBookingDetailBinding? = null
    private val binding get() = _binding!!
    private lateinit var viewModel: BookingDetailViewModel
    private var isCodeVisible = false
    private var rawBookingCode = ""

    @Inject
    lateinit var ticketCacheManager: TicketCacheManager

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentBookingDetailBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewModel = ViewModelProvider(this)[BookingDetailViewModel::class.java]

        val bookingId = arguments?.getString("bookingId")
        val paymentSuccess = arguments?.getBoolean("paymentSuccess", false) ?: false
        if (bookingId != null) {
            viewModel.loadBooking(bookingId)
        }

        // Nếu từ thanh toán thành công -> nút back chuyển sang tab Tickets
        if (paymentSuccess) {
            binding.btnBack.setOnClickListener { navigateToTicketsTab() }
            binding.btnBackTop.setOnClickListener { navigateToTicketsTab() }
        } else {
            binding.btnBack.setOnClickListener { Navigation.findNavController(view).popBackStack() }
            binding.btnBackTop.setOnClickListener { Navigation.findNavController(view).popBackStack() }
        }

        binding.ivToggleCode.setOnClickListener {
            isCodeVisible = !isCodeVisible
            updateBookingCodeDisplay()
        }

        binding.btnShare.setOnClickListener {
            if (rawBookingCode.isEmpty()) return@setOnClickListener
            val shareIntent = Intent(Intent.ACTION_SEND).apply {
                type = "text/plain"
                putExtra(Intent.EXTRA_SUBJECT, "Thông tin vé xem phim")
                val shareMessage = "Mình đã đặt vé xem phim *${binding.tvMovieTitle.text}*\n" +
                        "Tại: ${binding.tvCinema.text}\n" +
                        "Lúc: ${binding.tvShowtime.text}\n" +
                        "Mã đặt vé: $rawBookingCode"
                putExtra(Intent.EXTRA_TEXT, shareMessage)
            }
            startActivity(Intent.createChooser(shareIntent, "Chia sẻ vé"))
        }

        viewModel.booking.observe(viewLifecycleOwner) { resource ->
            when (resource.status) {
                Resource.Status.LOADING -> binding.progressBar.visibility = View.VISIBLE
                Resource.Status.SUCCESS -> {
                    binding.progressBar.visibility = View.GONE
                    resource.data?.let {
                        renderBooking(it, false)
                    }
                }
                Resource.Status.ERROR -> {
                    binding.progressBar.visibility = View.GONE
                    if (resource.data != null) {
                        renderBooking(resource.data, true)
                        if (!resource.message.isNullOrEmpty()) {
                            SnackbarHelper.showSuccess(binding.root, resource.message)
                        }
                    } else {
                        SnackbarHelper.showError(binding.root, resource.message ?: "Lỗi tải chi tiết vé")
                    }
                }
            }
        }
    }

    private fun renderBooking(b: BookingResponse, isOffline: Boolean) {
        rawBookingCode = b.bookingCode ?: ""
        updateBookingCodeDisplay()
        binding.tvMovieTitle.text = b.movieTitle
        binding.tvCinema.text = "CineNoir ${b.cinemaName ?: ""}"
        binding.tvScreen.text = b.screenName
        binding.tvFormat.text = formatScreenType(b.screenType)

        // Format showtime
        val rawTime = b.startTime
        if (rawTime != null) {
            try {
                val sdfIn = if (rawTime.contains("T")) {
                    SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.getDefault())
                } else {
                    SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault())
                }
                val d = sdfIn.parse(rawTime)
                if (d != null) {
                    val sdfOut = SimpleDateFormat("h:mm a - dd/MM/yyyy", Locale.getDefault())
                    binding.tvShowtime.text = sdfOut.format(d)
                }
            } catch (e: Exception) {
                binding.tvShowtime.text = rawTime
            }
        } else {
            binding.tvShowtime.text = ""
        }

        // Format status
        val statusText = when (b.status ?: "") {
            "PAID" -> "ĐÃ THANH TOÁN"
            "PENDING" -> "CHỜ THANH TOÁN"
            "CANCELLED" -> "ĐÃ HỦY"
            "EXPIRED" -> "HẾT HẠN"
            else -> b.status ?: ""
        }

        binding.tvStatus.text = statusText

        binding.tvTotal.text = String.format("Tổng thanh toán: %,.0fđ", b.totalAmount)

        // Danh sách ghế
        val seatsList = b.seats
        if (seatsList != null) {
            val seats = StringBuilder()
            for (i in seatsList.indices) {
                val s = seatsList[i]
                seats.append(s.rowLabel).append(s.colNumber)
                if (i < seatsList.size - 1) {
                    seats.append(", ")
                }
            }
            binding.tvSeats.text = seats.toString().trim()
        }

        // Danh sách bắp nước
        val combos = b.combos
        if (!combos.isNullOrEmpty()) {
            val combosStr = StringBuilder()
            for (k in combos.indices) {
                val cb = combos[k]
                combosStr.append(cb.quantity).append("x ").append(cb.comboName)
                if (k < combos.size - 1) {
                    combosStr.append("\n")
                }
            }
            binding.tvCombos.text = combosStr.toString().trim()
            binding.layoutCombos.visibility = View.VISIBLE
        } else {
            binding.layoutCombos.visibility = View.GONE
        }

        // Poster
        if (context != null) {
            Glide.with(this).load(b.moviePosterUrl)
                .placeholder(R.drawable.ic_movie_placeholder).into(binding.ivPoster)
        }

        // QR Code
        val qr = b.qrCode
        if (!qr.isNullOrEmpty()) {
            generateQrCode(qr)
        } else {
            binding.ivQrCode.visibility = View.GONE
        }

        // Trạng thái màu
        val statusColor = when (b.status ?: "") {
            "PAID" -> R.color.seat_available
            "PENDING" -> R.color.tertiary
            "CANCELLED", "EXPIRED" -> R.color.error
            else -> R.color.on_surface_variant
        }
        binding.tvStatus.setTextColor(resources.getColor(statusColor, null))

        if ("PENDING" == b.status && !isOffline) {
            binding.btnShare.visibility = View.GONE
            binding.btnPay.visibility = View.VISIBLE
            binding.btnPay.setOnClickListener {
                val args = Bundle().apply {
                    putString("bookingId", b.id)
                }
                if (view != null) {
                    Navigation.findNavController(requireView()).navigate(R.id.action_bookingDetail_to_payment, args)
                }
            }
        } else {
            binding.btnShare.visibility = View.VISIBLE
            binding.btnPay.visibility = View.GONE
        }
    }

    private fun updateBookingCodeDisplay() {
        if (_binding == null) return
        if (isCodeVisible) {
            binding.tvBookingCode.text = rawBookingCode
            binding.ivToggleCode.setImageResource(R.drawable.ic_eye)
            binding.tvBookingCode.letterSpacing = 0.3f
        } else {
            binding.tvBookingCode.text = "••••••"
            binding.ivToggleCode.setImageResource(R.drawable.ic_eye_off)
            binding.tvBookingCode.letterSpacing = 0.8f
        }
    }

    private fun generateQrCode(content: String) {
        try {
            val encoder = BarcodeEncoder()
            val bitmap = encoder.encodeBitmap(content, BarcodeFormat.QR_CODE, 400, 400)
            binding.ivQrCode.setImageBitmap(bitmap)
            binding.ivQrCode.visibility = View.VISIBLE
        } catch (e: Exception) {
            binding.ivQrCode.visibility = View.GONE
        }
    }

    private fun formatScreenType(rawType: String?): String {
        if (rawType == null) return "2D"
        return when (rawType.uppercase(Locale.getDefault())) {
            "STANDARD" -> "2D"
            "THREE_D" -> "3D"
            "IMAX" -> "IMAX"
            "FOUR_DX" -> "4DX"
            else -> rawType
        }
    }

    private fun navigateToTicketsTab() {
        try {
            val bottomNav = requireActivity().findViewById<BottomNavigationView>(R.id.bottomNav)
            if (bottomNav != null) {
                // Pop toàn bộ back stack về root trước
                Navigation.findNavController(requireView()).popBackStack(R.id.homeFragment, false)
                // Chuyển sang tab Tickets
                bottomNav.selectedItemId = R.id.bookingHistoryFragment
            }
        } catch (e: Exception) {
            // Fallback: quay về home
            Navigation.findNavController(requireView()).popBackStack(R.id.homeFragment, false)
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}
